main/src/ca/uwaterloo/flix/api/lsp/provider/HighlightProvider.scala

```diff
@@ -15,195 +16,369 @@
+    }
 
-        case Entity.Enum(enum0) => highlightEnum(uri, enum0.sym)
+    Visitor.visitRoot(root, LabelConsumer, acceptor)
 
-        case Entity.Struct(struct) => highlightStruct(uri, struct.sym)
+    builder.build
+  }
 
-        case Entity.TypeAlias(alias) => highlightTypeAlias(uri, alias.sym)
+  private def highlightTypeVarSym(sym: Symbol.KindedTypeVarSym)(implicit root: Root, acceptor: Visitor.Acceptor): JObject = {
+    val builder = HighlightBuilder(sym)
 
-        case Entity.AssocType(assoc) => highlightAssocType(uri, assoc.sym)
+    object TypeVarSymConsumer extends Consumer {
+      override def consumeTypeParam(tparam: TypedAst.TypeParam): Unit = builder.considerWrite(tparam.sym, tparam.sym.loc)
+      override def consumeType(tpe: Type): Unit = tpe match {
+        case Type.Var(sym, loc) => builder.considerRead(sym, loc)
+        case _ => ()
+      }
+    }
 
-        case Entity.Effect(eff) => highlightEffect(uri, eff.sym)
+    Visitor.visitRoot(root, TypeVarSymConsumer, acceptor)
 
-        case Entity.Op(op) => highlightOp(uri, op.sym)
+    builder.build
+  }
 
-        case Entity.VarUse(sym, _, _) => highlightVar(uri, sym)
+  private def highlightAssocTypeSym(sym: Symbol.AssocTypeSym)(implicit root: Root, acceptor: Visitor.Acceptor): JObject = {
+    val builder = HighlightBuilder(sym)
 
-        case Entity.DefUse(sym, _, _) => highlightDef(uri, sym)
+    object AssocTypeSymConsumer extends Consumer {
+      override def consumeAssocTypeSig(tsig: TypedAst.AssocTypeSig): Unit = builder.considerWrite(tsig.sym, tsig.sym.loc)
+      override def consumeAssocTypeSymUse(symUse: SymUse.AssocTypeSymUse): Unit = builder.considerRead(symUse.sym, symUse.loc)
+    }
 
-        case Entity.SigUse(sym, _, _) => highlightSig(uri, sym)
+    Visitor.visitRoot(root, AssocTypeSymConsumer, acceptor)
 
-        case Entity.CaseUse(sym, _, _) => highlightCase(uri, sym)
+    builder.build
+  }
+
+  /**
+    * A builder for creating an LSP highlight response containing a [[DocumentHighlight]] for each recorded
+    * occurrence of a given token [[tok]].
+    *
+    * An occurrence of a token of type [[T]] is "considered" by invoking [[considerRead]] or [[considerWrite]] for resp.
+    * "write" and "read" occurrences. By a "write" occurrence, we mean an occurrence where the token is being
+    * defined or otherwise bound to something. By "read" we mean an occurrence where the token is merely read.
+    * When we say "consider", we mean first checking if the occurrence is an occurrence of [[tok]] specifically.
+    * If so, it's added to our list of either "read" or "write" occurrences, depending on the type.
+    *
+    * When we're done considering tokens, we can construct the LSP response by calling [[build]]
+    *
+    * @param tok  the token we're finding occurrences of.
+    * @tparam T   the type of token that [[tok]] is.
+    */
+  private case class HighlightBuilder[T](tok: T) {
+    private var writes: List[SourceLocation] = Nil
```

Comment:
I think it is more natural to have reads before writes. So reorder everything below in this class according to that. Thanks!